File: Kotlin/ThreeNumbersSum.kt
package com.algoexpert.program
// O(n^2) time | O(n) space
fun threeNumberSum(array: MutableList<Int>, targetSum: Int): List<List<Int>> {
    array.sort()
	val triplets = mutableListOf<MutableList<Int>>()
	for (i in 0 until array.size - 2) {
		var left = i + 1
		var right = array.size - 1
		while (left < right) {
			val currentSum = array[i] + array[left] + array[right]
			if(currentSum == targetSum){
				triplets.add(mutableListOf(array[i], array[left], array[right]))
				left++
				right--
			} else if (currentSum < targetSum){
				left++
			} else if (currentSum > targetSum){
				right--
			}
		}
	}
    return triplets
}

File: Python/ThreeNumbersSum.py
# O(n^2) time | O(n) space
def threeNumbersSum(array, targetSum):
    array.sort()
    triplets = []
    for i in range(len(array) - 1):
        left = i + 1
        right = len(array) -1
        while left < right:
            currentSum = array[i] + array[left] + array[right]
            if currentSum == targetSum:
                triplets.append([array[i], array[left], array[right]])
                left += 1
                right -= 1
            elif currentSum < targetSum:
                left += 1
            elif currentSum > targetSum:
                right -= 1
    return triplets

nums = [1,3,4,5,8,-1,-4]

print(threeNumbersSum(nums, 9))